packages/integration-sdk-entity-validator/src/validator.ts

```diff
@@ -0,0 +1,158 @@
+    };
+  }
+
+  return {
+    schemaId: 'unknown',
+    message: 'Unknown error',
+    property: unknownPropertySymbol,
+    validation: 'unknown',
+  };
+};
+
+export class EntityValidator {
+  private ajvInstance: Ajv;
+
+  constructor({ schemas }: { schemas?: AnySchema[] }) {
+    this.ajvInstance = addJ1Formats(
+      addFormats(
+        new Ajv({
+          strictSchema: false,
+          allErrors: true,
+        }),
+      ),
+    );
+
+    if (schemas) {
+      this.addSchemas(schemas);
+    }
+  }
+
+  addSchemas(schema: AnySchema | AnySchema[]) {
+    this.ajvInstance.addSchema(schema);
+    return this;
+  }
+
+  validateEntity(
+    entity: object,
+    {
+      forceClassValidationWithValidatedType = false,
+    }: ValidateEntityOptions = {},
+  ): ValidateEntityResult {
+    const errors: EntityValidationError[] = [];
+    const skippedSchemas: SkippedSchema[] = [];
+
+    try {
+      assertEntity(entity);
+
+      const classArray = Array.isArray(entity._class)
+        ? entity._class
+        : [entity._class];
+
+      const typeSchemaId = `#${entity._type}`;
+      const typeValidator = this.ajvInstance.getSchema(typeSchemaId);
+
+      if (typeValidator) {
+        const isValid = typeValidator(entity);
+
+        if (!isValid) {
+          errors.push(
+            ...(typeValidator.errors?.map((error) =>
+              ajvErrorToEntityValidationError(typeSchemaId, error),
+            ) ?? []),
+          );
+        }
+      } else {
+        skippedSchemas.push({
+          schemaId: typeSchemaId,
+          reason: 'not-found',
+          type: 'type',
+        });
+      }
+
+      if (!typeValidator || forceClassValidationWithValidatedType) {
+        for (const className of classArray) {
+          const classSchemaId = `#${className}`;
+          const validator = this.ajvInstance.getSchema(classSchemaId);
+
+          if (!validator) {
+            skippedSchemas.push({
+              schemaId: classSchemaId,
+              reason: 'not-found',
+              type: 'class',
+            });
+            continue;
+          }
+
+          const isValid = validator(entity);
+
+          if (!isValid) {
+            errors.push(
+              ...(validator.errors?.map((error) =>
+                ajvErrorToEntityValidationError(classSchemaId, error),
+              ) ?? []),
+            );
+          }
+        }
+      } else {
+        skippedSchemas.push(
+          ...classArray.map(
+            (className) =>
+              ({
+                schemaId: `#${className}`,
+                reason: 'type-already-validated',
+                type: 'class',
+              }) satisfies SkippedSchema,
+          ),
+        );
+      }
+    } catch (error) {
+      errors.push(convertUnknownErrorToEntityValidationError(error));
+    }
+
+    return {
+      isValid: errors.length === 0,
+      errors: errors.length ? errors : null,
+      skippedSchemas: skippedSchemas.length ? skippedSchemas : null,
```

Comment:
We lost the information of if we used the `_type` vs `_class` for the validation. Is that deliberate?

Reply:
With the current implementation you could have both validations. You can get this information now from the skippedSchemas.

 Here is an example: 
![image](https://github.com/JupiterOne/sdk/assets/4943174/80ade7df-a8fa-4739-8faf-cdd5817f77c9)